### PR TITLE
DBSP projections

### PR DESCRIPTION
**Pull request by glommer**
This PR implements the ProjectOperator for DBSP circuits.
It is split in 3 commits, check the individual commits for details.

### CODE REVIEW

File: testing/materialized_views.test [hunk 1]
```diff
@@ -367,4 +367,21 @@ do_execsql_test_on_specific_db {:memory:} matview-mixed-operations-sequence {
...
+        SELECT b, a, b + a as c , (b * a) + 10 as d , min(a,b) as e
+        FROM t
+        where b > 2;
+
+    INSERT INTO t VALUES (1, 1);
+    INSERT INTO t VALUES (2, 2);
+    INSERT INTO t VALUES (3, 4);
+    INSERT INTO t VALUES (4, 3);
+
+    SELECT * from v;
```
Comment (jussisaurio) [inline testing/materialized_views.test:385]: Regarding the commit message about the previous tests having been wrong: there's a good opportunity here, while this is still quite early, to start differential fuzzing materialized views, and it's quite easy too:

```rust
assert_eq!(select * from my_matview, select * from (<my_matview_definition>)
```

We are missing a good "arbitrary sql generator" at the moment, but the plumbing can probably be lifted from some of the existing fuzz tests

File: core/incremental/operator.rs [hunk 2]
```diff
@@ -581,34 +582,189 @@ impl IncrementalOperator for FilterOperator {
...
+                            ) {
+                                columns.push(ProjectColumn {
+                                    expr,
+                                    alias: None,
+                                    compiled,
+                                });
+                            }
+                        }
+                    }
+                    _ => {
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: I think it would be great to improve on our previous bad decisions in tursodb in general and just return parse errors on any unsupported features. There's two big advantages to this:

1. It removes confusion from anyone trying to use the feature, and obviously makes returning incorrect results far less likely, and
2. It makes it easier for developers to see at a glance which features we still need to implement

File: core/incremental/operator.rs [hunk 3]
```diff
@@ -581,34 +582,189 @@ impl IncrementalOperator for FilterOperator {
...
+                        // For now, skip TableStar and other cases
+                    }
+                }
+            }
+
+            if columns.is_empty() {
+                // If no columns were extracted, default to projecting all input columns
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: Why?

File: core/incremental/operator.rs [hunk 4]
```diff
@@ -581,34 +582,189 @@ impl IncrementalOperator for FilterOperator {
...
+                            alias: None,
+                            compiled,
+                        })
+                    })
+                    .collect()
+            } else {
+                columns
+            }
+        } else {
+            // Not a simple SELECT statement, default to projecting all columns
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: Why?

File: core/incremental/operator.rs [hunk 5]
```diff
@@ -581,34 +582,189 @@ impl IncrementalOperator for FilterOperator {
...
+                            Ok(compiled) => {
+                                columns.push(ProjectColumn {
+                                    expr: expr.clone(),
+                                    alias: alias_str,
+                                    compiled,
+                                });
+                            }
+                            Err(_) => {
+                                // If compilation fails, skip this column for now
+                                // In the future we might want to handle this better
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: Why not now? Return a parse error

File: core/incremental/operator.rs [hunk 6]
```diff
@@ -581,34 +582,189 @@ impl IncrementalOperator for FilterOperator {
...
+                            }
+                        }
+                    }
+                    ResultColumn::Star => {
+                        // Select all columns - create trivial column references
+                        for name in &input_column_names {
+                            // Create an Id expression for the column
+                            let expr = Expr::Id(Name::Ident(name.clone()));
+                            // This should always compile successfully as a trivial column reference
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: Let's still handle the case where it doesn't

File: core/incremental/operator.rs [hunk 7]
```diff
@@ -581,34 +582,189 @@ impl IncrementalOperator for FilterOperator {
...
+                    .filter_map(|name| {
+                        let expr = Expr::Id(Name::Ident(name.clone()));
+                        CompiledExpression::compile(
+                            &expr,
+                            &input_column_names,
+                            schema,
+                            &temp_syms,
+                            internal_conn.clone(),
+                        )
+                        .ok()
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: let's handle the case where compilation fails

File: core/incremental/operator.rs [hunk 8]
```diff
@@ -581,34 +582,189 @@ impl IncrementalOperator for FilterOperator {
...
+                .filter_map(|name| {
+                    let expr = Expr::Id(Name::Ident(name.clone()));
+                    CompiledExpression::compile(
+                        &expr,
+                        &input_column_names,
+                        schema,
+                        &temp_syms,
+                        internal_conn.clone(),
+                    )
+                    .ok()
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: let's handle the case where compilation fails

File: core/incremental/operator.rs [hunk 9]
```diff
@@ -581,34 +582,189 @@ impl IncrementalOperator for FilterOperator {
...
-                ProjectColumn::Expression { alias, .. } => {
-                    alias.clone().unwrap_or_else(|| "expr".to_string())
-                }
+            .map(|c| {
+                c.alias.clone().unwrap_or_else(|| {
+                    // For simple column references, use the column name
+                    if let Expr::Id(name) = &c.expr {
+                        name.as_str().to_string()
+                    } else {
+                        "expr".to_string()
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: What about `Qualified`? Also, `Expr` implements `Display` so you could just call `expr.to_string()`

File: core/incremental/operator.rs [hunk 10]
```diff
@@ -620,24 +776,19 @@ impl ProjectOperator {
...
+            // Use the internal connection's pager for expression evaluation
+            let internal_pager = self.internal_conn.pager.borrow().clone();
+
+            // Execute the compiled expression (handles both columns and complex expressions)
+            let result = col
+                .compiled
+                .execute(values, internal_pager)
+                .unwrap_or_else(|_| {
+                    // Fall back to manual evaluation on error
+                    // This can happen for expressions with unsupported operations
```
Comment (jussisaurio) [inline core/incremental/operator.rs]: Let's just fail here then?

File: core/incremental/expr_compiler.rs [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+                            if *b != 0.0 {
+                                Value::Float(a / b)
+                            } else {
+                                Value::Null
+                            }
+                        }
+                        (Value::Null, _) | (_, Value::Null) => Value::Null,
+                        _ => panic!("Type mismatch in trivial expression: {left_val:?} / {right_val:?}. This is a bug in trivial expression validation."),
+                    },
+                    _ => panic!("Unsupported operator in trivial expression: {op:?}"),
```
Comment (jussisaurio) [inline core/incremental/expr_compiler.rs:162]: Good not to swallow the fallthrough case, but maybe make this function return a Result?

File: core/incremental/expr_compiler.rs [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+}
+
+impl CompiledExpression {
+    /// Get the "type" of a trivial expression for type checking
+    /// Returns None if type can't be determined statically
+    fn get_trivial_type(expr: &TrivialExpression) -> Option<&'static str> {
+        match expr {
+            TrivialExpression::Column(_) => None, // Can't know column type statically
+            TrivialExpression::Immediate(val) => match val {
+                Value::Integer(_) => Some("integer"),
```
Comment (jussisaurio) [inline core/incremental/expr_compiler.rs]: can this be an enum instead of str?

File: core/incremental/expr_compiler.rs [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+                } else {
+                    None // Type mismatch
+                }
+            }
+        }
+    }
+
+    /// Check if an expression is trivial (columns, immediates, and simple arithmetic)
+    /// Only considers expressions trivial if they don't require type coercion
+    fn is_trivial_expr(expr: &Expr, input_column_names: &[String]) -> Option<TrivialExpression> {
```
Comment (jussisaurio) [inline core/incremental/expr_compiler.rs]: `is_trivial_expr` suggests it returns a boolean, perhaps `try_get_trivial_expr()`?

File: core/incremental/expr_compiler.rs [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+                // Fast path: evaluate trivial expression inline
+                Ok(trivial.evaluate(values))
+            }
+            ExpressionExecutor::Compiled(program) => {
+                // Slow path: execute VDBE program
+                // Create a state with the input values loaded into registers
+                let mut state = ProgramState::new(program.max_registers, 0);
+
+                // Load input values into registers
+                for (idx, value) in values.iter().take(self.input_count).enumerate() {
```
Comment (jussisaurio) [inline core/incremental/expr_compiler.rs]: Should `values.len()` match `self.input_count`?

File: testing/materialized_views.test (hunk 1 above)
Comment (glommer) [reply testing/materialized_views.test:385]: I fully agree it has to be done, but I will focus mostly on getting the impl as far as I can for now

File: core/incremental/operator.rs (hunk 9 above)
Comment (glommer) [reply core/incremental/operator.rs]: And DoublyQualified too

File: core/incremental/operator.rs (hunk 10 above)
Comment (glommer) [reply core/incremental/operator.rs]: This one is the hardest one since project_values() and its callers do not return a Result. Part of that is because they will not be called at insert time, but rather asynchronously later when it is time to apply the delta into the accumulator. Perhaps this is a place for an assert?

**Comment by glommer** [comment]
@jussisaurio I addressed most of your comments.

File: core/incremental/expr_compiler.rs (hunk 11 above)
Comment (glommer) [reply core/incremental/expr_compiler.rs:162]: humm... The thing is that if this is not a trivial operation, then we should not have tried to construct a trivial operation out of it. This is just a violation about our assumptions about the code, which is usually what we use asserts for.

File: testing/materialized_views.test (hunk 1 above)
Comment (jussisaurio) [reply testing/materialized_views.test:385]: What is in scope for "the impl"? Our initial approach to testing in tursodb was that we supported the query translation / btree / vdbe modules basically with only handcrafted TCL tests and Rust unit tests, and once we finally started adding fuzz tests we discovered that, in every case, either 1. there were a ton of edge cases not covered or 2. the feature did not work at ALL (for example, it's fair to say that transactions did not - for all practical purposes - work at all when first released). Even today the fuzz coverage is bad because it's been retroactively bolted on instead of built with the feature set. So the earlier we do it the better.

Comment (glommer) [reply testing/materialized_views.test:385]: Personally, until we implement circuits and joins, I think we should be ready to throw away all of this code completely. The value of fuzz testing is then diminished.

Obviously I am hoping we are on the right track and will keep all of it.

**Comment by penberg** [comment]
Obsoleted by https://github.com/tursodatabase/turso/pull/2773
